core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java

```diff
@@ -0,0 +1,121 @@
+package eu.fasten.core.data.opal;
+
+import eu.fasten.core.data.opal.exceptions.MissingArtifactException;
+import eu.fasten.core.maven.utils.MavenUtilities;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.Arrays;
+import java.util.Optional;
+
+/**
+ * A set of methods for downloading POM and JAR files given Maven coordinates.
+ */
+public class MavenArtifactDownloader {
+    private static final Logger logger = LoggerFactory.getLogger(MavenArtifactDownloader.class);
+    private static final String[] packaging = {"jar", "war", "zip", "ear", "rar", "ejb", "par",
+            "aar", "car", "nar", "kar"};
+    private static final String[] defaultPackaging = {"zip", "aar", "tar.gz", "jar"};
+
+    /**
+     * Download a JAR file indicated by the provided Maven coordinate.
+     *
+     * @param mavenCoordinate A Maven coordinate in the for "groupId:artifactId:version"
+     * @return A temporary file on the filesystem
+     */
+    public File downloadArtifact(final MavenCoordinate mavenCoordinate, String artifactRepo)
```

Comment:
I would suggest to break this method down into several smaller helper methods... the complexity of the method goes through the roof!

Reply:
Addressed in 5f4cad7.